typings/index.d.ts

```diff
@@ -176,168 +176,9 @@ declare module 'discord.js' {
-      listener: (member: GuildMember | PartialGuildMember) => void,
-    ): this;
-    public once(
-      event: 'guildMembersChunk',
-      listener: (members: Collection<Snowflake, GuildMember | PartialGuildMember>, guild: Guild) => void,
-    ): this;
-    public once(
-      event: 'guildMemberSpeaking',
-      listener: (member: GuildMember | PartialGuildMember, speaking: Readonly<Speaking>) => void,
-    ): this;
-    public once(
-      event: 'guildMemberUpdate',
-      listener: (oldMember: GuildMember | PartialGuildMember, newMember: GuildMember | PartialGuildMember) => void,
-    ): this;
-    public once(event: 'guildUpdate', listener: (oldGuild: Guild, newGuild: Guild) => void): this;
-    public once(
-      event: 'message' | 'messageDelete' | 'messageReactionRemoveAll',
-      listener: (message: Message | PartialMessage) => void,
-    ): this;
-    public once(
-      event: 'messageDeleteBulk',
-      listener: (messages: Collection<Snowflake, Message | PartialMessage>) => void,
-    ): this;
-    public once(
-      event: 'messageReactionAdd' | 'messageReactionRemove',
-      listener: (messageReaction: MessageReaction, user: User | PartialUser) => void,
-    ): this;
-    public once(
-      event: 'messageUpdate',
-      listener: (oldMessage: Message | PartialMessage, newMessage: Message | PartialMessage) => void,
-    ): this;
-    public once(
-      event: 'presenceUpdate',
-      listener: (oldPresence: Presence | undefined, newPresence: Presence) => void,
-    ): this;
-    public once(event: 'rateLimit', listener: (rateLimitData: RateLimitData) => void): this;
-    public once(event: 'ready' | 'invalidated', listener: () => void): this;
-    public once(event: 'roleCreate' | 'roleDelete', listener: (role: Role) => void): this;
-    public once(event: 'roleUpdate', listener: (oldRole: Role, newRole: Role) => void): this;
-    public once(
-      event: 'typingStart',
-      listener: (channel: TextBasedChannelTypes | PartialChannel, user: User | PartialUser) => void,
-    ): this;
-    public once(
-      event: 'userUpdate',
-      listener: (oldUser: User | PartialUser, newUser: User | PartialUser) => void,
-    ): this;
-    public once(event: 'voiceStateUpdate', listener: (oldState: VoiceState, newState: VoiceState) => void): this;
-    public once(event: 'webhookUpdate', listener: (channel: TextChannel) => void): this;
-    public once(event: 'shardDisconnect', listener: (event: CloseEvent, id: number) => void): this;
-    public once(event: 'shardError', listener: (error: Error, id: number) => void): this;
-    public once(event: 'shardReady' | 'shardReconnecting', listener: (id: number) => void): this;
-    public once(event: 'shardResume', listener: (id: number, replayed: number) => void): this;
-    public once(event: string, listener: (...args: any[]) => void): this;
+    public once<K extends string>(event: K, listener: (...args: ClientEvents[K]) => void): this;
```

Comment:
To fix the issue mentioned by Sugden, the type parameter `T` should extend `keyof ClientEvents` instead of the string primitive.

Reply:
I agree that should be the type but it doesn't solve the IntelliSense issue, is it user friendly enough if we get rid of the last declaration in ClientEvents and let the user create custom events by extending the interface?
```typescript
// [k: string]: any[]; removed from ClientEvents

declare module "discord.js" {
  interface ClientEvents {
    myCustomEvent: [string];
  }
}

new Client().on("myCustomEvent", console.log);
```

Reply:
I feel that the benefit of IntelliSense auto completing the names more than makes up for this dip in usability. As far as I know, people who do this are the exception rather than the rule.